File: server/src/services/subAgentService.ts
import { v4 as uuidv4 } from 'uuid';
import { store } from '../db/memoryStore';
import type { SubAgentSession } from '../models/types';

/**
 * 启动一个 Sub-Agent 会话
 * - 创建会话，状态为 'running'
 * - 支持模拟流式输出
 */
export function startSubAgent(
  parentMessageId: string,
  task: string,
  context?: string,
  ownerId?: string,
): SubAgentSession {
  if (!parentMessageId || !task) {
    throw new Error('缺少必填字段: parentMessageId, task');
  }

  const now = Date.now();
  const session: SubAgentSession = {
    id: uuidv4(),
    parentMessageId,
    task,
    context: context?.trim() || undefined,
    status: 'running',
    output: '',
    createdAt: now,
    updatedAt: now,
  };

  store.subAgents.set(session.id, session);
  store.persistSubAgent(session);
  console.log(`[subagent] 启动子代理 ${session.id}, 任务: ${task.slice(0, 50)}`);
  return session;
}

/**
 * 向 Sub-Agent 发送人工输入
 * - 如果状态是 waiting_human，恢复为 running
 * - 将输入附加到上下文
 */
export function sendInput(
  subAgentId: string,
  input: string,
  _ownerId?: string,
): SubAgentSession {
  if (!subAgentId || !input) {
    throw new Error('缺少必填字段: subAgentId, input');
  }

  const session = store.subAgents.get(subAgentId);
  if (!session) {
    throw new Error('子代理会话不存在');
  }

  if (session.status === 'completed') {
    throw new Error('该子代理已完成，无法继续输入');
  }

  if (session.status === 'error') {
    throw new Error('该子代理出错，无法继续输入');
  }

  // 更新上下文
  session.context = (session.context ? session.context + '\n' : '') + `[Human]: ${input}`;
  session.status = 'running';
  session.updatedAt = Date.now();
  store.subAgents.set(subAgentId, session);
  store.persistSubAgent(session);

  console.log(`[subagent] 收到人工输入 ${subAgentId}`);
  return session;
}

/**
 * 中止 Sub-Agent 会话
 * - 设置状态为 'completed'
 */
export function abortSubAgent(
  subAgentId: string,
  _ownerId?: string,
): SubAgentSession {
  if (!subAgentId) {
    throw new Error('缺少必填字段: subAgentId');
  }

  const session = store.subAgents.get(subAgentId);
  if (!session) {
    throw new Error('子代理会话不存在');
  }

  if (session.status === 'completed' || session.status === 'error') {
    throw new Error('该子代理已结束');
  }

  session.status = 'completed';
  session.updatedAt = Date.now();
  store.subAgents.set(subAgentId, session);
  store.persistSubAgent(session);

  console.log(`[subagent] 中止子代理 ${subAgentId}`);
  return session;
}

/**
 * 获取 Sub-Agent 状态
 */
export function getSubAgentStatus(subAgentId: string): SubAgentSession {
  const session = store.subAgents.get(subAgentId);
  if (!session) {
    throw new Error('子代理会话不存在');
  }
  return session;
}

/**
 * 获取某个消息关联的所有 Sub-Agent 会话
 */
export function getSubAgentsForMessage(parentMessageId: string): SubAgentSession[] {
  const sessions: SubAgentSession[] = [];
  for (const session of store.subAgents.values()) {
    if (session.parentMessageId === parentMessageId) {
      sessions.push(session);
    }
  }
  // 按创建时间倒序
  sessions.sort((a, b) => b.createdAt - a.createdAt);
  return sessions;
}

File: server/services/sub_agent_service.py
"""
AICQ Sub-Agent Service
========================
Handles sub-agent lifecycle: starting, input provisioning, aborting,
and querying sub-agent status and results. Sub-agents are AI agents
spawned to handle sub-tasks within a parent message context.
"""

from __future__ import annotations

import logging
import uuid
from typing import Any, Dict, List, Optional

from db import DatabaseManager, json_serialize, iso_now
from config import Config

logger = logging.getLogger("aicq.sub_agent_service")


async def start_sub_agent(
    db: DatabaseManager,
    parent_message_id: str,
    task: str,
    context: Optional[str] = None,
    owner_id: Optional[str] = None,
) -> Dict[str, Any]:
    """Create a new sub-agent with status='running'.

    Args:
        parent_message_id: The ID of the parent message that spawned this sub-agent.
        task: The task description for the sub-agent.
        context: Optional context data for the sub-agent.
        owner_id: The account ID that owns this sub-agent.

    Returns the created sub-agent record.
    """
    sub_agent_id = uuid.uuid4().hex
    now = iso_now()

    await db.execute(
        """
        INSERT INTO sub_agents (id, parent_message_id, task, context, status, output, created_at, updated_at)
        VALUES (?, ?, ?, ?, 'running', '', ?, ?)
        """,
        (sub_agent_id, parent_message_id, task, context, now, now),
    )

    sub_agent = await get_sub_agent(db, sub_agent_id)
    logger.info("Sub-agent started: %s for message %s", sub_agent_id, parent_message_id)
    return sub_agent


async def send_input(
    db: DatabaseManager,
    sub_agent_id: str,
    input_data: str,
) -> Dict[str, Any]:
    """Provide input to a sub-agent that is waiting for human input.

    Updates status from 'waiting_human' to 'running' and appends the
    input data to the context.

    Raises ValueError if the sub-agent is not found or not in 'waiting_human' status.
    """
    sub_agent = await db.fetchone("SELECT * FROM sub_agents WHERE id = ?", (sub_agent_id,))
    if not sub_agent:
        raise ValueError("Sub-agent not found")

    if sub_agent["status"] != "waiting_human":
        raise ValueError(f"Sub-agent is not waiting for input (status: {sub_agent['status']})")

    now = iso_now()

    # Append input to existing context
    existing_context = sub_agent.get("context") or ""
    updated_context = existing_context + "\n" + input_data if existing_context else input_data

    await db.execute(
        """
        UPDATE sub_agents
        SET status = 'running', context = ?, updated_at = ?
        WHERE id = ?
        """,
        (updated_context, now, sub_agent_id),
    )

    sub_agent = await get_sub_agent(db, sub_agent_id)
    logger.info("Input sent to sub-agent: %s", sub_agent_id)
    return sub_agent


async def abort_sub_agent(
    db: DatabaseManager,
    sub_agent_id: str,
) -> Dict[str, Any]:
    """Abort a sub-agent.

    Sets status to 'error' with an abort message in the output.

    Raises ValueError if the sub-agent is not found or already completed.
    """
    sub_agent = await db.fetchone("SELECT * FROM sub_agents WHERE id = ?", (sub_agent_id,))
    if not sub_agent:
        raise ValueError("Sub-agent not found")

    if sub_agent["status"] in ("completed", "error"):
        raise ValueError(f"Cannot abort sub-agent in '{sub_agent['status']}' status")

    now = iso_now()
    await db.execute(
        """
        UPDATE sub_agents
        SET status = 'error', output = 'Aborted by user', updated_at = ?
        WHERE id = ?
        """,
        (now, sub_agent_id),
    )

    sub_agent = await get_sub_agent(db, sub_agent_id)
    logger.info("Sub-agent aborted: %s", sub_agent_id)
    return sub_agent


async def get_sub_agent(
    db: DatabaseManager,
    sub_agent_id: str,
) -> Optional[Dict[str, Any]]:
    """Get sub-agent info by ID. Returns None if not found."""
    sub_agent = await db.fetchone("SELECT * FROM sub_agents WHERE id = ?", (sub_agent_id,))
    return sub_agent


async def get_sub_agents_by_message(
    db: DatabaseManager,
    parent_message_id: str,
) -> List[Dict[str, Any]]:
    """List all sub-agents for a given parent message."""
    sub_agents = await db.fetchall(
        "SELECT * FROM sub_agents WHERE parent_message_id = ? ORDER BY created_at ASC",
        (parent_message_id,),
    )
    return sub_agents
